src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs

```diff
@@ -2618,6 +2618,101 @@ internal static int IndexOfAnyValueType<T>(ref T searchSpace, T value0, T value1
             return -1;
         }
 
+        public static void Replace<T>(Span<T> span, T oldValue, T newValue)
+        {
+            for (int i = 0; i < span.Length; ++i)
+            {
+                ref T val = ref span[i];
+
+                if (EqualityComparer<T>.Default.Equals(oldValue, val))
+                {
+                    val = newValue;
+                }
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void ReplaceValueType<T>(ref T src, ref T dst, T oldValue, T newValue, nuint length) where T : struct
+        {
+            Debug.Assert(Vector128.IsHardwareAccelerated && Vector128<T>.IsSupported);
+
+            if (length < (uint)Vector128<T>.Count)
+            {
+                for (nuint idx = 0; idx < length; ++idx)
+                {
+                    T original = Unsafe.Add(ref src, idx);
+                    Unsafe.Add(ref dst, idx) = EqualityComparer<T>.Default.Equals(original, oldValue) ? newValue : original;
+                }
+            }
+            else
+            {
+                ReplaceValueTypeVectorized(ref src, ref dst, oldValue, newValue, length);
+            }
+        }
+
+        private static void ReplaceValueTypeVectorized<T>(ref T src, ref T dst, T oldValue, T newValue, nuint length) where T : struct
+        {
+            Debug.Assert(Vector128.IsHardwareAccelerated && Vector128<T>.IsSupported);
+
+            nuint idx = 0;
+
+            if (!Vector256.IsHardwareAccelerated || length < (uint)Vector256<T>.Count)
+            {
+                nuint lastVectorIndex = length - (uint)Vector128<T>.Count;
+                Vector128<T> oldValues = Vector128.Create(oldValue);
+                Vector128<T> newValues = Vector128.Create(newValue);
+
+                do
+                {
+                    Vector128<T> original = Vector128.LoadUnsafe(ref src, idx);
+                    Vector128<T> mask = Vector128.Equals(oldValues, original);
+                    Vector128<T> result = Vector128.ConditionalSelect(mask, newValues, original);
+                    result.StoreUnsafe(ref dst, idx);
+
+                    idx += (uint)Vector128<T>.Count;
+                } while (idx <= lastVectorIndex);
+
+                if (idx < length)
```

Comment:
Previously you argued against such a guard check.  You measured and changed your mind?

Reply:
This is really hard to measure, as in micro-benchmarks the branch-predictor will make good decisions.
I'm glad you picked this one up, as I just have my view on this and would like to see some other thoughts or a confirmation.

My take is as follows:
[Previously](https://github.com/dotnet/runtime/pull/67049#discussion_r946253469) we had `Vector<ushort>`, so on AVX that's 16 elements. Thus the probability that no remainder needs to be done is $\frac{1}{16}$ (6.25 %), or in other words: with the guard we penalize $\frac{15}{16}$ (93.75 %) of inputs as the remainders needs to be processed.
Additionally the vector read should be in the cache (as just read a few instructions before, leaving context-switching aside), and an additional store should be handled by cpu's store buffers.

Now we have `Vector128<T>` / `Vector256<T>`, so the probability of not having a remainder changes depending on the size of T.
| size | elements Vec128 | p for remainder Vec 128 | elements Vec256 | p for remainder Vec 256 |
| -- | -- | -- | -- | -- |
| 1 | 16 | $\frac{15}{16}$ (93.75 %) | 32 | $\frac{31}{32}$ (96.875 %) |
| 2 | 8 | $\frac{7}{8}$ (87.5 %) | 16 | $\frac{14}{16}$ (93.75 %) |
| 4 | 4 | $\frac{3}{4}$ (75 %) | 8 | $\frac{7}{8}$ (87.5 %) |
| 8 | 2 | $\frac{1}{2}$ (50 %) | 4 | $\frac{3}{4}$ (75 %) |

So based on that one can argue that the branch-predictor could naively always take the branch (i.e. process the remainder) as this is the most likely case.
Thus the guard is just overhead (fetch, decode, bigger code-size).

In the rare case that the remainder doesn't need to be processed, and the branch-predictor took the branch a pipeline stall will need to happen, costing about 20 cycles.
Just executing the code cost 13 cycles according [godbolt](https://godbolt.org/z/b48G85Tbd).

Sure this is rather naive approach to the branch-predictor, but based on that I'm leaning towards removing that guard for both Vector128 and Vector256, i.e. process the remainder anyway as it's cheaper based on that theory.

<sub>Note: this is of course no rule of thumb, applies only to this use case.</sub>

Reply:
Ey, in the whole handling of remainders (of idempotent) operations I thought quite a bit, but not far enough...(well mentally caught in the history and scheme that vec-loops _must_ process as much as possible)
Sometimes it can be so easy to just remove a character from the code 😉 
-> ebcbb9fa1e25b0661b150f562b3b398039cac13e (as seen in other recent PRs)